### DISCUSSION TITLE
A simple case of Compensable workflow

### DISCUSSION THREAD

**Discussion by herme063** [elsa-workflows/elsa-core#4280] [Q&A]
I have a programmatic workflow:

	builder
		.WriteLine("Let's get started...")
		.Then<Compensable>(comp => 
		{
			comp.When(OutcomeNames.Body)
				.Then<Step1>()
				.Then<Step2>();
				
			comp.When(OutcomeNames.Compensate)
				.Then<Comp>();
		});

I return `Fault("")` in **Step1** activity but **Comp** activity was not called.

I do the same for **Step2** and it behaves the same.

My question is, how do I compensate for any fault/cancellation occurring in the Body section?

**Comment by sfmskywalker** [comment]
Hello @herme063,

Thanks for reaching out with your question!

Here's how compensation works in Elsa 2:

- The Compensate branch is executed **only** if a fault happens after the Compensable's Done outcome.
- If a fault takes place in the Body branch of the Compensable, no compensation will occur.

To achieve compensation, you can modify your workflow as follows:

```csharp
builder
	.WriteLine("Let's get started...")
	.Then<Compensable>(comp => 
	{
		comp.When(OutcomeNames.Compensate)
			.Then<Comp>();
	})
        .Then<Step1>() // If this faults, the Compensate branch will be triggered next, and Step 2 won't execute.
	.Then<Step2>();
```

I did notice an issue when testing the above, which was due to a bug in the engine. I've addressed this, so ensure you're using the latest preview packages or have pulled the most recent source code if you're running from there.

You might be curious about the purpose of the Body branch. Honestly, I've overlooked its intended use, but I'll revisit it when working on compensation in Elsa 3.

For another compensable workflow example, you can refer to [this example](https://github.com/elsa-workflows/elsa-core/blob/master/src/samples/console/Elsa.Samples.CompensationConsole/Workflows/CompensableWorkflow.cs).

**Comment by herme063** [comment]
Thank you!
